File: Programmers/C++/메뉴 리뉴얼.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <unordered_map>
#define FOR(i,n) for(int i=0;i<(n);++i)
#define FORS(i,s,n) for(int i=(s);i<(n);++i)

using namespace std;

int mx;
vector<string> res;
unordered_map<string, int> um;

void dfs(int cnt, int idx, string ori, string s) {
    if(cnt == s.size()) {
        um[s]++;
        return;
    }
    FORS(i,idx,ori.size()) {
        s+=ori[i];
        dfs(cnt,i+1,ori,s);
        s.pop_back();
    }
}

vector<string> solution(vector<string> orders, vector<int> course) {
    for(int cnt: course) {
        mx=0;
        um.clear();
        string tmp = "";
        for(string s: orders) {
            sort(s.begin(),s.end());
            tmp = "";
            dfs(cnt,0,s,tmp);
        }
        for(auto k: um) mx=max(mx,k.second);
        for(auto k: um) {
            if(k.second==mx && mx>=2) res.push_back(k.first);
        }
    }
    sort(res.begin(),res.end());
    return res;
}

File: Programmers/Swift/메뉴 리뉴얼.swift
import Foundation

var res = [String]()
var hash = [[String:Int]](repeating: [String:Int](), count: 11)

func dfs(_ idx:Int, _ cnt:Int, _ s:[Character], _ cur: inout [Character]) {
    if cnt == cur.count {
        var str = cur.map{String($0)}.joined()
        var val = 0
        if var num = hash[cnt][str] {val=num}
        hash[cnt].updateValue(val+1, forKey: str)
        return
    }
    for i in idx..<s.count {
        cur.append(s[i])
        dfs(i+1,cnt,s,&cur)
        cur.removeLast()
    }
}

func solution(_ orders:[String], _ course:[Int]) -> [String] {
    
    for i in 0..<orders.count {
        var s = Array(orders[i])
        s.sort()
        for cnt in 0...s.count {
            var cur = [Character]()
            dfs(0,cnt,s,&cur)
        }
    }
    
    for num in course {
        var mx = 0
        for (key, value) in hash[num] {
            mx = max(mx, value)
        }
        if mx <= 1 {continue}
        for (key, value) in hash[num] {
            if mx == value {res.append(key)}
        }
    }
    
    res.sort()
    
    return res
}
